samza-core/src/main/java/org/apache/samza/zk/ZkBarrierForVersionUpgrade.java

```diff
@@ -117,41 +132,50 @@ public void join(String version, String participantId) {
+    }
   }
+
   /**
    * Listener for changes to the list of participants. It is meant to be subscribed only by the creator of the barrier
    * node. It checks to see when the barrier is ready to be marked as completed.
    */
   class ZkBarrierChangeHandler extends ZkUtils.GenIZkChildListener {
     private final String barrierVersion;
-    private final List<String> names;
+    private final List<String> expectedParticipantIds;
 
-    public ZkBarrierChangeHandler(String barrierVersion, List<String> names, ZkUtils zkUtils) {
+    public ZkBarrierChangeHandler(String barrierVersion, List<String> expectedParticipantIds, ZkUtils zkUtils) {
       super(zkUtils, "ZkBarrierChangeHandler");
       this.barrierVersion = barrierVersion;
-      this.names = names;
+      this.expectedParticipantIds = expectedParticipantIds;
     }
 
     @Override
-    public void handleChildChange(String parentPath, List<String> currentChildren) {
+    public void handleChildChange(String barrierParticipantPath, List<String> participantIds) {
       if (notAValidEvent()) {
         return;
       }
-      if (currentChildren == null) {
-        LOG.info("Got ZkBarrierChangeHandler handleChildChange with null currentChildren");
+      if (participantIds == null) {
+        LOG.info("Received notification with null participants for barrier: {}. Ignoring it.", barrierParticipantPath);
         return;
       }
-      LOG.info("list of children in the barrier = " + parentPath + ":" + Arrays.toString(currentChildren.toArray()));
-      LOG.info("list of children to compare against = " + parentPath + ":" + Arrays.toString(names.toArray()));
+      LOG.info(String.format("Current participants in barrier version: %s = %s.", barrierVersion, participantIds));
+      LOG.info(String.format("Expected participants in barrier version: %s = %s.", barrierVersion, expectedParticipantIds));
 
       // check if all the expected participants are in
-      if (currentChildren.size() == names.size() && CollectionUtils.containsAll(currentChildren, names)) {
-        String barrierDonePath = keyBuilder.getBarrierStatePath(barrierVersion);
-        LOG.info("Writing BARRIER DONE to " + barrierDonePath);
-        zkUtils.writeData(barrierDonePath, State.DONE); // this will trigger notifications
-        zkUtils.unsubscribeChildChanges(barrierDonePath, this);
+      if (participantIds.size() == expectedParticipantIds.size() && CollectionUtils.containsAll(participantIds, expectedParticipantIds)) {
+        String barrierStatePath = keyBuilder.getBarrierStatePath(barrierVersion);
+        State barrierState = zkUtils.getZkClient().readData(barrierStatePath);
+        if (Objects.equals(barrierState, State.NEW)) {
+          LOG.info(String.format("Expected participants has joined the barrier version: %s. Marking the barrier state: %s as %s.", barrierVersion, barrierStatePath, State.DONE));
+          zkUtils.writeData(barrierStatePath, State.DONE); // this will trigger notifications
```

Comment:
Not necessarily in this PR but would be good to think about what would happen if the barrier state changed in the time we do this check and act on it.

Reply:
exactly what I meant, thanks!